core/sr-io/src/offchain/http.rs

```diff
@@ -0,0 +1,398 @@
+			})
+			.collect()
+	}
+}
+
+/// A HTTP response.
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Response {
+	/// Request id
+	pub id: RequestId,
+	/// Response status code
+	pub code: u16,
+	/// A collection of headers.
+	headers: Option<Headers>,
+}
+
+impl Response {
+	fn new(id: RequestId, code: u16) -> Self {
+		Self {
+			id,
+			code,
+			headers: None,
+		}
+	}
+
+	/// Retrieve the headers for this response.
+	pub fn headers(&mut self) -> &Headers {
+		if self.headers.is_none() {
+			self.headers = Some(Headers { raw: crate::http_response_headers(self.id) });
+		}
+		self.headers.as_ref().expect("Headers were just set; qed")
+	}
+
+	/// Retrieve the body of this response.
+	pub fn body(&self) -> ResponseBody {
+		ResponseBody::new(self.id)
+	}
+}
+
+/// A buffered byte iterator over response body.
+pub struct ResponseBody {
+	id: RequestId,
+	is_deadline_reached: bool,
+	buffer: [u8; 4096],
+	filled_up_to: Option<usize>,
+	position: usize,
+	deadline: Option<Timestamp>,
+}
+
+#[cfg(feature = "std")]
+impl std::fmt::Debug for ResponseBody {
+	fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+		fmt.debug_struct("ResponseBody")
+			.field("id", &self.id)
+			.field("is_deadline_reached", &self.is_deadline_reached)
+			.field("buffer", &self.buffer.len())
+			.field("filled_up_to", &self.filled_up_to)
+			.field("position", &self.position)
+			.field("deadline", &self.deadline)
+			.finish()
+	}
+}
+
+impl ResponseBody {
+	fn new(id: RequestId) -> Self {
+		ResponseBody {
+			id,
+			is_deadline_reached: false,
+			buffer: [0_u8; 4096],
+			filled_up_to: None,
+			position: 0,
+			deadline: None,
+		}
+	}
+
+	/// Set the deadline for reading the body.
+	pub fn deadline(&mut self, deadline: impl Into<Option<Timestamp>>) {
+		self.deadline = deadline.into();
+	}
+
+	/// Check if the iterator has finished because of reaching the deadline.
+	pub fn is_deadline_reached(&self) -> bool {
+		self.is_deadline_reached
+	}
+}
+
+impl Iterator for ResponseBody {
+	type Item = u8;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		if self.is_deadline_reached {
+			return None;
+		}
+
+		if self.filled_up_to.is_none() {
+			let result = crate::http_response_read_body(self.id, &mut self.buffer, self.deadline).ok();
+			if let Some(size) = result {
+				self.position = 0;
+				self.filled_up_to = Some(size);
+			} else {
+				self.is_deadline_reached = true;
+				return None;
+			}
+		}
+
+		if Some(self.position) == self.filled_up_to {
+			self.filled_up_to = None;
+			return self.next();
+		}
+
+		let result = self.buffer[self.position];
+		self.position += 1;
+		Some(result)
+	}
+}
+
+/// A collection of Headers in the response.
+#[derive(Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Headers {
+	/// Raw headers
+	pub raw: Vec<(Vec<u8>, Vec<u8>)>,
+}
+
+impl Headers {
+	/// Retrieve a single header from the list of headers.
+	///
+	/// Note this method is linearly looking from all the headers.
```

Comment:
doesn't `from_utf8` make it quadratic?

Reply:
Ah you're right, mb sth like this

```suggestion
	/// Note this method is linearly looking from all the headers comparing them with the needle byte-by-byte.
```